internal/db/embedding.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2025 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package db
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/philippgille/chromem-go"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/errors"
+	"github.com/sourcenetwork/defradb/internal/keys"
+)
+
+func getEmbeddingFunc(provider, model, url string) (chromem.EmbeddingFunc, error) {
+	switch provider {
+	case "ollama":
+		return chromem.NewEmbeddingFuncOllama(model, url), nil
+	case "openai":
+		normalized := true
+		apiURL := url
+		if apiURL == "" {
+			apiURL = chromem.BaseURLOpenAI
+		}
+		return chromem.NewEmbeddingFuncOpenAICompat(
+			apiURL,
+			os.Getenv("OPENAI_API_KEY"),
+			model,
+			&normalized,
+		), nil
+	default:
+		return nil, NewErrUnknownEmbeddingProvider(provider)
+	}
+}
+
+// setEmbedding sets the embedding fields on the document if the related fields are dirty.
+// However, if the vector field itself has been set, it will not be overwritten by a new embedding generation.
+func (c *collection) setEmbedding(ctx context.Context, doc *client.Document, isCreate bool) error {
+	embeddingGenerated := false
+	for _, embedding := range c.Description().Embeddings {
+		vecValue, err := doc.GetValue(embedding.FieldName)
+		if err != nil && !errors.Is(err, client.ErrFieldNotExist) {
+			return NewErrGetEmbeddingField(err)
+		}
+		if vecValue != nil && vecValue.IsDirty() {
+			// If the vector has been explicitly set, no need to generate the embedding.
+			continue
```

Comment:
suggestion: This line shows untested, worth investigating why, if actually missing coverage, then add test please.

Reply:
CI issue